src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionEscapeHelper.ts

```diff
@@ -0,0 +1,25 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { GeneralShellType, TerminalShellType, WindowsShellType } from '../../../../../platform/terminal/common/terminal.js';
+
+/**
+ * Escapes special characters in a file/folder label for shell completion.
+ * This ensures that characters like [, ], etc. are properly escaped.
+ */
+export function escapeTerminalCompletionLabel(label: string, shellType: TerminalShellType | undefined, pathSeparator: string): string {
+	// Only escape for bash/zsh/fish; PowerShell and cmd have different rules
+	if (!shellType || shellType === WindowsShellType.CommandPrompt || shellType === GeneralShellType.PowerShell) {
+		return label;
+	}
+
+	// Bash/zsh/fish escaping
+	const specialCharsSet = new Set(['[', ']', '(', ')', '{', '}', `'`, '"', '\\', '$', '`', ';', '&', '|', '<', '>', '*', '?', '~', '#', '=', '%', '!']);
+	const specialCharsRegex = /[\[\]\(\)\{\}'"\\$\`;&|<>*?~#=%!]/;
+	if (!specialCharsRegex.test(label)) {
+		return label;
+	}
+	return label.split('').map(c => specialCharsSet.has(c) ? '\\' + c : c).join('');
+}
```

Comment:
I don't think we can do this generally, these characters may be perfectly valid in the context and this would likely introduce a host of other problems.

The safest approach here is to escape on the completion item in the provider, so escape for the git branch generator and for paths when we know it's safe to do. We also want the label to represent exactly what will be typed and this changes that.

Reply:
ok, re 
> We also want the label to represent exactly what will be typed and this changes that.

That's how I initially implemented it. Then changed the approach because it looked a bit ugly 

<img width="568" alt="Screenshot 2025-06-20 at 9 28 20 AM" src="https://github.com/user-attachments/assets/fe89bc0c-e83b-4b17-a45c-7486dad23d72" />

Reply:
went back there

Reply:
we resolve the paths in core, not the extension, so cannot do this for just git specifically. What is a case when this is not safe or desirable?

Reply:
For git it would need to be in the post processing of the generator:

https://github.com/microsoft/vscode/blob/315b0949786b3807f05cb6acd13bf0029690a052/extensions/terminal-suggest/src/completions/upstream/git.ts#L54

Reply:
I think we should do this for more than git